rust/src/mqtt/parser.rs

```diff
@@ -327,125 +341,151 @@ pub fn parse_msgidonly_v3(i: &[u8]) -> IResult<&[u8], MQTTMessageIdOnly> {
 }
 
 #[inline]
-pub fn parse_subscribe_topic(i: &[u8]) -> IResult<&[u8], MQTTSubscribeTopicData> {
+fn parse_subscribe_topic(i: &[u8]) -> IResult<&[u8], MQTTSubscribeTopicData> {
     let (i, topic_name) = parse_mqtt_string(i)?;
     let (i, qos) = be_u8(i)?;
     Ok((i, MQTTSubscribeTopicData { topic_name, qos }))
 }
 
-pub fn parse_subscribe(i: &[u8], protocol_version: u8) -> IResult<&[u8], MQTTSubscribeData> {
-    let (i, message_id) = be_u16(i)?;
-    let (i, properties) = parse_properties(i, protocol_version == 5)?;
-    let (i, topics) = many1(complete(parse_subscribe_topic))(i)?;
-    Ok((
-        i,
-        MQTTSubscribeData {
-            message_id,
-            topics,
-            properties,
-        },
-    ))
+#[inline]
+fn parse_subscribe(protocol_version: u8) -> impl Fn(&[u8]) -> IResult<&[u8], MQTTSubscribeData>
+where
+{
+    move |i: &[u8]| {
+        let (i, message_id) = be_u16(i)?;
+        let (i, properties) = parse_properties(i, protocol_version == 5)?;
+        let (i, topics) = many1(complete(parse_subscribe_topic))(i)?;
+        Ok((
+            i,
+            MQTTSubscribeData {
+                message_id,
+                topics,
+                properties,
+            },
+        ))
+    }
 }
 
-pub fn parse_suback(i: &[u8], protocol_version: u8) -> IResult<&[u8], MQTTSubackData> {
-    let (i, message_id) = be_u16(i)?;
-    let (qoss, properties) = parse_properties(i, protocol_version == 5)?;
-    Ok((
-        i,
-        MQTTSubackData {
-            message_id,
-            qoss: qoss.to_vec(),
-            properties,
-        },
-    ))
+#[inline]
+fn parse_suback(protocol_version: u8) -> impl Fn(&[u8]) -> IResult<&[u8], MQTTSubackData>
+where
+{
+    move |i: &[u8]| {
+        let (i, message_id) = be_u16(i)?;
+        let (qoss, properties) = parse_properties(i, protocol_version == 5)?;
+        Ok((
+            i,
+            MQTTSubackData {
+                message_id,
+                qoss: qoss.to_vec(),
+                properties,
+            },
+        ))
+    }
 }
 
-pub fn parse_unsubscribe(i: &[u8], protocol_version: u8) -> IResult<&[u8], MQTTUnsubscribeData> {
-    let (i, message_id) = be_u16(i)?;
-    let (i, properties) = parse_properties(i, protocol_version == 5)?;
-    let (i, topics) = many0(complete(parse_mqtt_string))(i)?;
-    Ok((
-        i,
-        MQTTUnsubscribeData {
-            message_id,
-            topics,
-            properties,
-        },
-    ))
+#[inline]
+fn parse_unsubscribe(
+    protocol_version: u8,
+) -> impl Fn(&[u8]) -> IResult<&[u8], MQTTUnsubscribeData>
+where
```

Comment:
Is the `where` needed here?

Reply:
It looks like it's not needed in these cases so I can drop it from the code if you want. Will do that in the next PR.